In [2]:
system("ls", intern=T)

[1] "bioinfo1.ipynb"                  "drosophila_promoters_500_100.fa"
[3] "human gene analysis.ipynb"       "lab2.ipynb"                     
[5] "lesson4.ipynb"                   "mart_export.txt"                
[7] "notes"                           "sample_promoters.fa"            
[9] "word counts.ipynb"

In [4]:
library("Biostrings")
library("BSgenome")

Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs

The following objects are masked from ‘package:base’:

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which, which.max, which.min

Loading required package: S4Vectors
Loading required package: stats4

Attaching package: ‘S4Vectors’

The followin

In [7]:
#import sequences

seq <- readDNAStringSet("dmelanogaster_promoter_100_to_0.fa")

In [22]:
myString <- toString(seq[[1]])

In [25]:
for (index in 1:(length(seq p1)-5)) {
    print()
}



[1] NA
[1] NA
[1] NA
[1] NA
[1] NA
[1] "CTCTTAAAATTTGTGTTTTGGAGAGACTCGTTCTCATCTCTCGTTGAGCGACGCGAGCGAGAGAGGCAAACTATTTAAGCCACATCGCAAGGCAACTCGT"


In [31]:
seq[1]


  A DNAStringSet instance of length 1
    width seq                                               names               
[1]   100 CTCTTAAAATTTGTGTTTTGGAG...AGCCACATCGCAAGGCAACTCGT FBgn0000008

In [224]:
?extractAt #to retrieve words from sequences

In [118]:
#indexing of a DNAStringSet requires an IRange index
at1 <- IRanges(start = 1:95, width = 6)

#extraction
words <- extractAt(seq, at1)

In [226]:
words[1]

DNAStringSetList of length 1
[["FBgn0000008"]] CTCTTA TCTTAA CTTAAA TTAAAA ... GCAACT CAACTC AACTCG ACTCGT

In [169]:
#convert to data.frame, doesn't work.
words_list <- as.data.frame(words)
(words_list)

group group_name  value 
1       1     FBgn0000008 CTCTTA
2       1     FBgn0000008 TCTTAA
3       1     FBgn0000008 CTTAAA
4       1     FBgn0000008 TTAAAA
5       1     FBgn0000008 TAAAAT
6       1     FBgn0000008 AAAATT
7       1     FBgn0000008 AAATTT
8       1     FBgn0000008 AATTTG
9       1     FBgn0000008 ATTTGT
10      1     FBgn0000008 TTTGTG
11      1     FBgn0000008 TTGTGT
12      1     FBgn0000008 TGTGTT
13      1     FBgn0000008 GTGTTT
14      1     FBgn0000008 TGTTTT
15      1     FBgn0000008 GTTTTG
16      1     FBgn0000008 TTTTGG
17      1     FBgn0000008 TTTGGA
18      1     FBgn0000008 TTGGAG
19      1     FBgn0000008 TGGAGA
20      1     FBgn0000008 GGAGAG
21      1     FBgn0000008 GAGAGA
22      1     FBgn0000008 AGAGAC
23      1     FBgn0000008 GAGACT
24      1     FBgn0000008 AGACTC
25      1     FBgn0000008 GACTCG
26      1     FBgn0000008 ACTCGT
27      1     FBgn0000008 CTCGTT
28      1     FBgn0000008 TCGTTC
29      1     FBgn0000008 CGTTCT
30      1     FBgn0000008 GTTCTC
⋮       ⋮     ⋮           ⋮     
1323416 13931 FBgn0286222 GGTCCG
1323417 13931 FBgn0286222 GTCCGA
1323418 13931 FBgn0286222 TCCGAT
1323419 13931 FBgn0286222 CCGATA
1323420 13931 FBgn0286222 CGATAG
1323421 13931 FBgn0286222 GATAGC
1323422 13931 FBgn0286222 ATAGCT
1323423 13931 FBgn0286222 TAGCTC
1323424 13931 FBgn0286222 AGCTCT
1323425 13931 FBgn0286222 GCTCTC
1323426 13931 FBgn0286222 CTCTCA
1323427 13931 FBgn0286222 TCTCAC
1323428 13931 FBgn0286222 CTCACC
1323429 13931 FBgn0286222 TCACCC
1323430 13931 FBgn0286222 CACCCG
1323431 13931 FBgn0286222 ACCCGC
1323432 13931 FBgn0286222 CCCGCG
1323433 13931 FBgn0286222 CCGCGC
1323434 13931 FBgn0286222 CGCGCA
1323435 13931 FBgn0286222 GCGCAT
1323436 13931 FBgn0286222 CGCATC
1323437 13931 FBgn0286222 GCATCG
1323438 13931 FBgn0286222 CATCGA
1323439 13931 FBgn0286222 ATCGAG
1323440 13931 FBgn0286222 TCGAGG
1323441 13931 FBgn0286222 CGAGGT
1323442 13931 FBgn0286222 GAGGTG
1323443 13931 FBgn0286222 AGGTGC
1323444 13931 FBgn0286222 GGTGCT
1323445 13931 FBgn0286222 GTGCTG

In [216]:
# initialize and empty dataframe, set the columns
df <- as.data.frame(matrix(nrow=0, ncol=96))
names(df) = c("gene", 1:95)
names(df)

[1] "gene" "1"    "2"    "3"    "4"    "5"    "6"    "7"    "8"    "9"   
[11] "10"   "11"   "12"   "13"   "14"   "15"   "16"   "17"   "18"   "19"  
[21] "20"   "21"   "22"   "23"   "24"   "25"   "26"   "27"   "28"   "29"  
[31] "30"   "31"   "32"   "33"   "34"   "35"   "36"   "37"   "38"   "39"  
[41] "40"   "41"   "42"   "43"   "44"   "45"   "46"   "47"   "48"   "49"  
[51] "50"   "51"   "52"   "53"   "54"   "55"   "56"   "57"   "58"   "59"  
[61] "60"   "61"   "62"   "63"   "64"   "65"   "66"   "67"   "68"   "69"  
[71] "70"   "71"   "72"   "73"   "74"   "75"   "76"   "77"   "78"   "79"  
[81] "80"   "81"   "82"   "83"   "84"   "85"   "86"   "87"   "88"   "89"  
[91] "90"   "91"   "92"   "93"   "94"   "95"

In [357]:
#populate the dataframe

genesList <- names(words)

for (g in genesList) {  # "FBgn0286222"
    
    df[g,] <- c(g, as.list(as.data.frame(words[[g]]))$x)
}

#head(df)

gene        1      2      3      4      5      6      7      8     
FBgn0000008 FBgn0000008 CTCTTA TCTTAA CTTAAA TTAAAA TAAAAT AAAATT AAATTT AATTTG
FBgn0000015 FBgn0000015 GTCGGC TCGGCG CGGCGT GGCGTC GCGTCG CGTCGT GTCGTA TCGTAT
FBgn0000014 FBgn0000014 CCGCCG CGCCGC GCCGCT CCGCTG CGCTGC GCTGCT CTGCTG TGCTGC
FBgn0000028 FBgn0000028 AGCGGG GCGGGC CGGGCA GGGCAG GGCAGG GCAGGA CAGGAC AGGACA
FBgn0000018 FBgn0000018 AAAAGC AAAGCA AAGCAT AGCATT GCATTA CATTAA ATTAAT TTAATA
FBgn0000017 FBgn0000017 CTGCCA TGCCAA GCCAAC CCAACG CAACGC AACGCG ACGCGT CGCGTT
            9      ⋯ 86     87     88     89     90     91     92     93    
FBgn0000008 ATTTGT ⋯ CGCAAG GCAAGG CAAGGC AAGGCA AGGCAA GGCAAC GCAACT CAACTC
FBgn0000015 CGTATA ⋯ AACTAA ACTAAC CTAACA TAACAG AACAGC ACAGCG CAGCGA AGCGAT
FBgn0000014 GCTGCA ⋯ CTCGCA TCGCAT CGCATG GCATGG CATGGC ATGGCT TGGCTT GGCTTT
FBgn0000028 GGACAC ⋯ ATCATA TCATAG CATAGA ATAGAA TAGAAT AGAATT GAATTC AATTCT
FBgn0000018 TAATAT ⋯ CGGTCA GGTCAC GTCACA TCACAC CACACT ACACTT CACTTT ACTTTA
FBgn0000017 GCGTTC ⋯ GACCGA ACCGAC CCGACA CGACAA GACAAG ACAAGC CAAGCG AAGCGG
            94     95    
FBgn0000008 AACTCG ACTCGT
FBgn0000015 GCGATT CGATTC
FBgn0000014 GCTTTT CTTTTG
FBgn0000028 ATTCTT TTCTTT
FBgn0000018 CTTTAC TTTACT
FBgn0000017 AGCGGT GCGGTC

In [358]:
df.backup <- df

In [337]:
df <- df.backup

In [359]:
df$gene <- NULL

In [360]:
dfdt<-as.data.table(df)

In [361]:
#library(purrr)

In [362]:
#todo: identify the 15 most frequent words in our dataset
#todo: find where they are located in our sequences

#1) find word counts, concatenating the dataframe and calling table()
wordcount <- table(do.call(c,dfdt))

In [363]:
#find different words
wordcount <- sort(wordcount, decreasing = T)

In [364]:
topwords <- wordcount[1:15]

In [365]:
#library(tidyverse)

In [366]:
index_words <- names(wordcount)[1]
columns <- c(1:95)

we can use `stringi::stri_rand_shuffle` to randomly permute letters in a word. This is to chech that our top words are really important, and not just due to the higher composition of their nucleotides.

In [344]:
regexSearch = "TATAAA"

In [353]:
dfdt[is.element(dfdt, "TATAAA")]

ERROR: Error in `[.data.table`(dfdt, is.element(dfdt, "TATAAA")): i evaluates to a logical vector length 96 but there are 13931 rows. Recycling of logical i is no longer allowed as it hides more bugs than is worth the rare convenience. Explicitly use rep(...,length=.N) if you really need to recycle.


To check if a word is significant, we can perform tests. A simple test are binomial test, Fisher test, etc.
Our universe is the total number of words in our dataset. `nrows(df)`= 13931.

We also need the total number of occurrencies for every word.

In [354]:
wordcount


     TTTTTT      AAAAAA      ATCGAT      TATTTT      ATTTTT      ATAAAA 
       3017        2806        2601        2582        2315        2240 
     TATAAA      TTTAAA      TCGATA      AAAATA      ATATTT      AAAAAT 
       2208        2205        2120        2110        2096        2049 
     TAAAAA      TATCGA      TTAAAA      AAATAT      TTATTT      TTTTAA 
       2035        1964        1937        1854        1819        1818 
     TTTTTA      AAATAA      ATAAAT      ATATAT      TATATA      ATTTAA 
       1813        1720        1658        1658        1655        1654 
     TAAATA      AAATTT      AATAAA      TAAAAT      TTTATT      AATTTT 
       1650        1648        1645        1603        1584        1562 
     AAAATT      ATTTTA      TTAAAT      TATTTA      TTTTTG      TTTTAT 
       1559        1555        1546        1541        1474        1454 
     ATATAA      GTATTT      TATATT      ATTTTC      TAAATT      AATATT 
       1397        1385        1375        1371   

In [356]:
names(wordcount)

[1] "TTTTTT"      "AAAAAA"      "ATCGAT"      "TATTTT"      "ATTTTT"     
    [6] "ATAAAA"      "TATAAA"      "TTTAAA"      "TCGATA"      "AAAATA"     
   [11] "ATATTT"      "AAAAAT"      "TAAAAA"      "TATCGA"      "TTAAAA"     
   [16] "AAATAT"      "TTATTT"      "TTTTAA"      "TTTTTA"      "AAATAA"     
   [21] "ATAAAT"      "ATATAT"      "TATATA"      "ATTTAA"      "TAAATA"     
   [26] "AAATTT"      "AATAAA"      "TAAAAT"      "TTTATT"      "AATTTT"     
   [31] "AAAATT"      "ATTTTA"      "TTAAAT"      "TATTTA"      "TTTTTG"     
   [36] "TTTTAT"      "ATATAA"      "GTATTT"      "TATATT"      "ATTTTC"     
   [41] "TAAATT"      "AATATT"      "AAAAAC"      "TTTTTC"      "AAAACA"     
   [46] "TTTTCA"      "ATTTAT"      "CAAAAA"      "TAATTT"      "ATTAAA"     
   [51] "AAATTA"      "AAACAA"      "AATTTA"      "TGTTTT"      "AAAAGC"     
   [56] "CATTTT"      "GTTTTT"      "TTGTTT"      "AACAAA"      "AATTAA"     
   [61] "TTCAAA"      "AATATA"      "TTTAAT"      "TTAATT"      "TTTTGT"     
   [66] "TTTGAA"      "ATTATT"      "TTTGTT"      "CGATAG"      "AATCGA"     
   [71] "TTTCAA"      "TAAAAG"      "AAAATG"      "ATTTTG"      "GAAAAA"     
   [76] "AAAAAG"      "CGATAA"      "CAGCTG"      "AATAAT"      "TGAAAA"     
   [81] "ACTTTT"      "TTGAAA"      "GAAAAT"      "ACAAAA"      "AAAAGT"     
   [86] "TCGATT"      "TTTTGA"      "ACATTT"      "CTATCG"      "CTTTTT"     
   [91] "TTTTCT"      "TTTATA"      "AAATAC"      "TTTTGG"      "AAATGT"     
   [96] "TCAAAA"      "CAAAAT"      "TTATCG"      "AGTTTT"      "ATTTCA"     
  [101] "AAAATC"      "TATTTC"      "TTTTCG"      "TTTTCC"      "AAAACT"     
  [106] "ATAATT"      "GATTTT"      "GCTTTT"      "AAACAT"      "AATTAT"     
  [111] "CAAATT"      "TCAAAT"      "TGAAAT"      "AAATTC"      "TTTCGA"     
  [116] "TTATAT"      "AAATTG"      "AATTTC"      "AATTTG"      "TATTAA"     
  [121] "TTTCTT"      "ATAATA"      "TTATTA"      "GAAATT"      "TTTTGC"     
  [126] "TAATAA"      "AGAAAA"      "TCAGTT"      "ATTAAT"      "ATTTGT"     
  [131] "CGATAT"      "TATTTG"      "TTATAA"      "TTCAGT"      "TTCGAT"     
  [136] "CAATTT"      "TCATTT"      "TGGTAT"      "CTCTCT"      "AAATCA"     
  [141] "TATTAT"      "AGCTTT"      "GTATAT"      "ATCGAA"      "AAAGCT"     
  [146] "CGAAAA"      "TAATTA"      "TTCATT"      "GGAAAA"      "TCGAAA"     
  [151] "TTTTAG"      "TTTCAT"      "TTTCCA"      "TGGCAA"      "TAAAAC"     
  [156] "ATTCAA"      "GTTTTC"      "GCATTT"      "AACATT"      "CCAAAA"     
  [161] "TAATAT"      "GAATTT"      "ATTTGA"      "ATATCG"      "TTTGGC"     
  [166] "TTGCCA"      "ATATTA"      "TTAATA"      "CTTTTC"      "AACTTT"     
  [171] "ACAAAT"      "GTATAA"      "TGATTT"      "AAGAAA"      "AACAGC"     
  [176] "AAAGTT"      "TTCTTT"      "AATTCA"      "TCTCTC"      "GGTATT"     
  [181] "ATTTCT"      "CGATTT"      "TAAACA"      "CATCGA"      "GTTTTG"     
  [186] "TTGAAT"      "AATCAA"      "GAAAAC"      "CAAATA"      "ATGTTT"     
  [191] "TTTCAG"      "CTTTTG"      "TAAATG"      "TTCAAT"      "TGGAAA"     
  [196] "CTATAA"      "CATTTA"      "TGAATT"      "GAAATA"      "AAATCG"     
  [201] "ATTTCG"      "AAATGC"      "AAATGA"      "GATAAG"      "GCTGTT"     
  [206] "AAACTT"      "ATGAAA"      "TATAAT"      "AACAAT"      "AAAAGA"     
  [211] "TTTGGT"      "AATGTT"      "TTTGTA"      "AATGAA"      "AAGTTT"     
  [216] "ACATAT"      "TTTAAG"      "CTAAAA"      "AAATAG"      "TCAATT"     
  [221] "GCAAAA"      "TGTGTG"      "ATCAAA"      "ATAACA"      "CAAACA"     
  [226] "TTTAGT"      "CATTTC"      "GGCAGC"      "AAAGTA"      "AGCAAA"     
  [231] "AAAGCA"      "TACAAA"      "CATTTG"      "TGGCCA"      "AAAGAA"     
  [236] "TTTGCT"      "AATATC"      "TGTAAA"      "TGTTTA"      "AAAACC"     
  [241] "CAAAAC"      "TCTTTT"      "CAGTTT"      "GTTTTA"      "CGTTTT"     
  [246] "CTTTTA"      "AGAGAG"      "ATTGAA"      "TTTTAC"      "AATTGT"     
  [251] "TACTTT"      "ATTGTT"      "TTTGCA"      "ATAAAC"      "GCTGCT"     
  [256] "TGCTGC"      "CAA

check if proportion of occurrencies of our found frequencies are different from the frequencies of random strings can be performed with 

In [355]:
observed = #freq from wordcounts
random   = #freq from random words
total    = #total number of seqs: 1323445

prop.test(c(observed, random), c(total,total))

ERROR: Error in is.table(x): object 'observed' not found


In [377]:
#library(stringi)
# randomized sequences

randomSeq <- sapply(seq, stri_rand_shuffle)
